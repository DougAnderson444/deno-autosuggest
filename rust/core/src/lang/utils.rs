use std::collections::HashMap;


pub fn compile_utf_map(static_map: &[(&str, &str)]) -> HashMap<Vec<char>, Vec<char>> {
    let mut compiled_map = HashMap::with_capacity(static_map.len());
    for (pattern, replace) in static_map {
        compiled_map.insert(
            pattern.chars().collect(),
            replace.chars().collect()
        );
    }
    compiled_map
}
