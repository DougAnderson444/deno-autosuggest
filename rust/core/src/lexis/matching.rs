use std::fmt;
use std::collections::HashSet;
use crate::damlev::DamerauLevenshtein;
use crate::jaccard::Jaccard;
use super::{Word, Text};


const LENGTH_THRESHOLD:  f64 = 0.51;
const JACCARD_THRESHOLD: f64 = 0.41;
const DAMLEV_THRESHOLD:  f64 = 0.21;

thread_local! {
    static DAMLEV:  DamerauLevenshtein<char> = DamerauLevenshtein::new();
    static JACCARD: Jaccard<char>            = Jaccard::new();
}


#[derive(Clone, PartialEq)]
pub struct WordMatch {
    pub query:  MatchSide,
    pub record: MatchSide,
    pub typos:  usize,
    pub fin:    bool,
}


#[derive(Clone, PartialEq, Debug)]
pub struct MatchSide {
    pub ix:      usize,
    pub len:     usize,
    pub slice:   (usize, usize),
    pub primary: bool,
}


impl MatchSide {
    pub fn new(len: usize, slice: (usize, usize), primary: bool) -> Self {
        MatchSide { ix: 0, len, slice, primary }
    }
}


impl fmt::Debug for WordMatch {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "WordMatch {{ ")?;

        for i in 0 .. self.record.len {
            if i == self.record.slice.0 { write!(f, "[")?; }
            write!(f, "r")?;
            if i + 1 == self.record.slice.1 { write!(f, "]")?; }
        }

        write!(f, " /{}/ ", self.typos)?;

        for i in 0 .. self.query.len {
            if i == self.query.slice.0 { write!(f, "[")?; }
            write!(f, "q")?;
            if i + 1 == self.query.slice.1 { write!(f, "]")?; }
        }
        if !self.fin {
            write!(f, "..")?;
        }

        write!(f, " }}")?;
        Ok(())
    }
}


pub fn text_match(rtext: &Text<&[char]>, qtext: &Text<&[char]>) -> Vec<WordMatch> {
    let capacity = min!(rtext.words.len(), qtext.words.len());
    let mut taken: HashSet<usize> = HashSet::with_capacity(capacity);
    let mut matches: Vec<WordMatch> = Vec::with_capacity(capacity);

    for (i, qword) in qtext.words.iter().enumerate() {
        let mut found: Option<WordMatch> = None;

        for (j, rword) in rtext.words.iter().enumerate() {
            if taken.contains(&j) { continue; }
            if let Some(mut m) = word_match(rword, qword, &rtext.chars, &qtext.chars) {
                m.query.ix  = i;
                m.record.ix = j;
                if found.is_none() && !m.record.primary {
                    found = Some(m);
                    continue;
                }
                if m.record.primary {
                    found = Some(m);
                    break;
                }
            }
        }

        if let Some(m) = found {
            taken.insert(m.record.ix);
            matches.push(m);
        }
    }

    matches
}


pub fn word_match(rword: &Word, qword: &Word, rchars: &[char], qchars: &[char]) -> Option<WordMatch> {
    if qword.is_empty() || rword.is_empty() {
        return None;
    }
    if !length_check(rword, qword) {
        return None;
    }
    if !jaccard_check(rword, qword, rchars, qchars) {
        return None;
    }

    let rchars = rword.view(rchars);
    let qchars = qword.view(qchars);
    let mut best_match = None;

    DAMLEV.with(|damlev| {
        damlev.distance(qchars, rchars);
        let dists  = &*damlev.dists.borrow();

        let left  = if qword.fin { max!(qword.stem, rword.stem) } else { qword.stem } - 1;
        let right = max!(qword.len(), rword.len()) + 1;

        if right <= left { return; }

        let range = (left .. right).rev();

        for rlen in range.clone() {
            for qlen in range.clone() {
                // Out of bounds.
                if qlen > qword.len() { continue; }
                if rlen > rword.len() { continue; }
                // Left margin is for insertion/deletion, not for both prefixes at the same time.
                if rlen == left  && qlen == left  { continue; }
                // Compare full words only if query is finished.
                if qword.fin && rlen < rword.stem { break; }
                if qword.fin && qlen < qword.stem { break; }
                // Words with 2+ insertions/deletions are mismatched by default.
                if (qlen as isize - rlen as isize).abs() > 1 { continue; }

                let dist = dists.get(qlen + 1, rlen + 1);
                let rel  = dist as f64 / max!(qlen, rlen, 1) as f64;

                if rel > DAMLEV_THRESHOLD { continue; }

                match best_match {
                    None => {
                        best_match = Some(WordMatch {
                            query:  MatchSide::new(qlen,        (0, qlen), qword.is_primary()),
                            record: MatchSide::new(rword.len(), (0, rlen), rword.is_primary()),
                            typos:  dist,
                            fin:    qword.fin || rword.len() == rlen,
                        });
                    },
                    Some(ref mut m) => {
                        if m.typos <= dist { continue; }
                        m.record.slice = (0, rlen);
                        m.typos        = dist;
                        m.fin          = qword.fin || rword.len() == rlen;
                    },
                }

                if dist == 0 {
                    break;
                }
            }
        }
    });

    best_match
}


pub fn length_check(rword: &Word, qword: &Word) -> bool {
    let qlen  = qword.len();
    let rlen  = if qword.fin { rword.len() } else { min!(qlen, rword.len()) };

    if qlen <= 3 || rlen <= 3 {
        return qlen == rlen;
    }

    let long  = max!(qlen, rlen);
    let short = min!(qlen, rlen);
    let dist  = 1.0 - (short as f64 / long as f64);

    dist < LENGTH_THRESHOLD
}


pub fn jaccard_check(rword: &Word, qword: &Word, rchars: &[char], qchars: &[char]) -> bool {
    let rchars = rword.view(rchars);
    let qchars = qword.view(qchars);
    let rslice = if qword.fin { &rchars } else { &rchars[.. min!(qword.len(), rword.len())] };
    let dist   = JACCARD.with(|j| j.rel_dist(&rslice, &qchars));
    dist < JACCARD_THRESHOLD
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::lexis::Chars;
    use crate::lang::lang_english;
    use super::{Word, Text, word_match, text_match};


    fn text(s: &str) -> Text<Vec<char>> {
        Text::from_str(s).split(&Chars::Whitespaces)
    }


    // Match word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_empty_both() {
        let qchars = "".chars().collect::<Vec<_>>();
        let rchars = "".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_eq!(word_match(&rword, &qword, &rchars[..], &qchars[..]), None);
    }


    #[test]
    fn match_word_empty_record() {
        let qchars = "mailbox".chars().collect::<Vec<_>>();
        let rchars = "".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_eq!(word_match(&rword, &qword, &rchars[..], &qchars[..]), None);
    }


    #[test]
    fn match_word_empty_query() {
        let qchars = "".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_eq!(word_match(&rword, &qword, &rchars[..], &qchars[..]), None);
    }


    // Match word: prefixed by same length word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_full_strict() {
        let qchars = "mailbox".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_fuzzy_insertion() {
        let qchars = "mailybox".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_fuzzy_deletion() {
        let qchars = "mailox".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_fuzzy_transposition() {
        let qchars = "maiblox".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_query_too_long() {
        let qchars1 = "mailboxes".chars().collect::<Vec<_>>();
        let qchars2 = "mailboxes".chars().collect::<Vec<_>>();
        let rchars  = "mail"     .chars().collect::<Vec<_>>();
        let qword1  = Word::new(qchars1.len()).fin(true);
        let qword2  = Word::new(qchars2.len()).fin(false);
        let rword   = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword1, &rchars[..], &qchars1[..]));
        assert_debug_snapshot!(word_match(&rword, &qword2, &rchars[..], &qchars2[..]));
    }

    #[test]
    fn match_word_full_stem() {
        let     rchars  = "universe"  .chars().collect::<Vec<_>>();
        let     qchars1 = "university".chars().collect::<Vec<_>>();
        let     qchars2 = "university".chars().collect::<Vec<_>>();
        let mut rword   = Word::new(rchars.len());
        let mut qword1  = Word::new(qchars1.len());
        let     qword2  = Word::new(qchars2.len());

        let lang = lang_english();
        qword1.stem(&qchars1[..], &lang);
        rword .stem(&rchars[..],  &lang);

        assert_debug_snapshot!(word_match(&rword, &qword1, &rchars[..], &qchars1[..]));
        assert_debug_snapshot!(word_match(&rword, &qword2, &rchars[..], &qchars2[..]));
    }


    // Match word: prefixed by lesser length word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_partial_strict() {
        let qchars = "mailb".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword = Word::new(qchars.len()).fin(false);
        let rword = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_partial_fuzzy_insertion() {
        let qchars = "maiylb".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_partial_fuzzy_deletion() {
        let qchars = "maib".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_partial_fuzzy_transposition() {
        let qchars = "malib".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    // Match text
    // ----------------------------------------------------------------

    #[test]
    fn match_text_empty_both() {
        let qtext = text("").fin(false);
        let rtext = text("");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_empty_one() {
        let qtext = text("mailbox").fin(false);
        let rtext = text("");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
        assert_debug_snapshot!(text_match(&qtext.to_ref(), &rtext.to_ref()));
    }


    #[test]
    fn match_text_singleton_equality() {
        let qtext = text("mailbox").fin(false);
        let rtext = text("mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_singleton_typos() {
        let qtext = text("maiblox").fin(false);
        let rtext = text("mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_first() {
        let qtext = text("yelow").fin(false);
        let rtext = text("yellow mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_second() {
        let qtext = text("maiblox").fin(false);
        let rtext = text("yellow mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_unfinished() {
        let qtext = text("maiblox yel").fin(false);
        let rtext = text("yellow mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_intersection() {
        let qtext = text("big malibox yelo").fin(false);
        let rtext = text("small yellow metal mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_best_rword() {
        let mut qtext = text("the").fin(false);
        let mut rtext = text("the theme");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));

        let lang = lang_english();
        qtext = qtext.mark_pos(&lang);
        rtext = rtext.mark_pos(&lang);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_regression_best_match() {
        let qtext = text("sneak").fin(false);
        let rtext = text("sneaky");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }
}