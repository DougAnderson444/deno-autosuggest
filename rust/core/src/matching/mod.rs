mod damlev;
mod jaccard;
mod word;
mod text;

use std::fmt;
use crate::tokenization::{Word};
pub use word::word_match;
pub use text::text_match;


#[derive(Clone, PartialEq, Debug)]
pub struct WordMatch {
    pub query:  MatchSide,
    pub record: MatchSide,
    pub typos:  usize,
    pub fin:    bool,
}


impl WordMatch {
    pub fn new(
        qword: &Word,
        rword: &Word,
        qlen:  usize,
        rlen:  usize,
        typos: usize
    ) -> Self {
        Self {
            query: MatchSide {
                ix:      qword.ix,
                len:     qword.len(),
                slice:   (0, qlen),
                primary: qword.is_primary(),
            },
            record: MatchSide {
                ix:      rword.ix,
                len:     rword.len(),
                slice:   (0, rlen),
                primary: rword.is_primary(),
            },
            typos,
            fin: qword.fin || rword.len() == rlen,
        }
    }

    pub fn split_query(&self, w1: &Word, w2: &Word) -> (Self, Self) {
        let (query1, query2) = self.query.split(w1, w2);
        let typos2 = self.typos / 2;
        let typos1 = self.typos - typos2;
        let part1  = Self {
            query:  query1,
            record: self.record.clone(),
            typos:  typos1,
            fin:    true,
        };
        let part2 = Self {
            query:  query2,
            record: self.record.clone(),
            typos:  typos2,
            fin:    self.fin,
        };
        (part1, part2)
    }

    pub fn split_record(&self, w1: &Word, w2: &Word) -> (Self, Self) {
        let (record1, record2) = self.record.split(w1, w2);
        let typos2 = self.typos / 2;
        let typos1 = self.typos - typos2;
        let part1  = Self {
            query:  self.query.clone(),
            record: record1,
            typos:  typos1,
            fin:    true,
        };
        let part2 = Self {
            query:  self.query.clone(),
            record: record2,
            typos:  typos2,
            fin:    self.fin,
        };
        (part1, part2)
    }
}


#[derive(Clone, PartialEq, Debug)]
pub struct MatchSide {
    pub ix:      usize,
    pub len:     usize,
    pub slice:   (usize, usize),
    pub primary: bool,
}

impl MatchSide {
    pub fn split(&self, w1: &Word, w2: &Word) -> (Self, Self) {
        let part1  = Self {
            ix:      w1.ix,
            len:     w1.len(),
            slice:   (0, w1.len()),
            primary: w1.is_primary(),
        };
        let part2 = Self {
            ix:      w2.ix,
            len:     w2.len(),
            slice:   (0, self.slice.1 - (w2.place.0 - w1.place.0)),
            primary: w2.is_primary(),
        };
        (part1, part2)
    }
}


impl fmt::Display for WordMatch {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "WordMatch {{ ")?;

        for i in 0 .. self.record.len {
            if i == self.record.slice.0 { write!(f, "[")?; }
            write!(f, "r")?;
            if i + 1 == self.record.slice.1 { write!(f, "]")?; }
        }

        write!(f, " /{}/ ", self.typos)?;

        for i in 0 .. self.query.len {
            if i == self.query.slice.0 { write!(f, "[")?; }
            write!(f, "q")?;
            if i + 1 == self.query.slice.1 { write!(f, "]")?; }
        }
        if !self.fin {
            write!(f, "..")?;
        }

        write!(f, " }}")?;
        Ok(())
    }
}
