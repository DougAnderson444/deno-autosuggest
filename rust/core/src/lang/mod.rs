mod utils;
mod normalize;
mod lang_english;
mod lang_german;
mod lang_portuguese;
mod lang_russian;
mod lang_spanish;

pub use lang_german::lang_german;
pub use lang_english::lang_english;
pub use lang_spanish::lang_spanish;
pub use lang_portuguese::lang_portuguese;
pub use lang_russian::lang_russian;

use std::cell::RefCell;
use std::collections::HashMap;
use rust_stemmers::Stemmer;
use crate::tokenization::PartOfSpeech;
use normalize::Normalize;


const BUFFER_CAPACITY: usize = 2;


pub struct Lang {
    stem_buffer:  RefCell<String>,
    norm_buffer1: RefCell<Vec<char>>,
    norm_buffer2: RefCell<Vec<char>>,
    pos_map:      HashMap<Vec<char>, PartOfSpeech>,
    compose_map:  HashMap<Vec<char>, Vec<char>>,
    reduce_map:   HashMap<Vec<char>, Vec<char>>,
    stemmer:      Stemmer,
}


impl Lang {
    pub fn new(
        pos_map:     HashMap<Vec<char>, PartOfSpeech>,
        compose_map: HashMap<Vec<char>, Vec<char>>,
        reduce_map:  HashMap<Vec<char>, Vec<char>>,
        stemmer:     Stemmer
    ) -> Self {
        let stem_buffer  = RefCell::new(String::with_capacity(BUFFER_CAPACITY));
        let norm_buffer1 = RefCell::new(Vec::with_capacity(BUFFER_CAPACITY));
        let norm_buffer2 = RefCell::new(Vec::with_capacity(BUFFER_CAPACITY));
        Self { stem_buffer, norm_buffer1, norm_buffer2, pos_map, compose_map, reduce_map, stemmer }
    }

    pub fn stem(&self, word: &[char]) -> usize {
        let buffer = &mut *self.stem_buffer.borrow_mut();
        buffer.clear();
        buffer.extend(word.iter());
        let stem = self.stemmer.stem(&buffer);
        stem.chars().count()
    }

    pub fn get_pos(&self, word: &[char]) -> Option<PartOfSpeech> {
        self.pos_map.get(word).cloned()
    }

    pub fn utf_compose(&self, word: &[char]) -> Option<Vec<char>> {
        let buffer = &mut *self.norm_buffer1.borrow_mut();
        buffer.clear();

        for (_, norm_chunk) in Normalize::new(word, &self.compose_map) {
            buffer.extend(norm_chunk);
        }

        if &buffer[..] == word {
            None
        } else {
            Some(buffer.clone())
        }
    }

    pub fn utf_reduce(&self, word: &[char]) -> Option<(Vec<char>, Vec<char>)> {
        let buffer1 = &mut *self.norm_buffer1.borrow_mut();
        let buffer2 = &mut *self.norm_buffer2.borrow_mut();
        buffer1.clear();
        buffer2.clear();

        for (word_chunk, norm_chunk) in Normalize::new(word, &self.reduce_map) {
            buffer1.extend(word_chunk);
            buffer2.extend(norm_chunk);
            for _ in 0 .. norm_chunk.len() - word_chunk.len() {
                buffer1.push('\0');
            }
        }

        if &buffer2[..] == word {
            None
        } else {
            Some((buffer1.clone(), buffer2.clone()))
        }
    }
}


#[cfg(test)]
mod tests {
    use std::collections::HashMap;
    use rust_stemmers::{Stemmer, Algorithm};
    use insta::assert_debug_snapshot;
    use super::Lang;
    use super::utils::compile_utf_map;

    fn get_lang() -> Lang {
        let stemmer  = Stemmer::create(Algorithm::English);
        let compose_map = compile_utf_map(&[("ó", "ó")]);
        let reduce_map  = compile_utf_map(&[
            ("ó", "o"),
            ("ß", "ss"),
        ]);
        let lang = Lang::new(HashMap::new(), compose_map, reduce_map, stemmer);
        lang
    }

    #[test]
    fn utf_compose_nfc() {
        let input  = "foóbar".chars().collect::<Vec<_>>();
        let output = get_lang().utf_compose(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_compose_nfd() {
        let input  = "foóbar".chars().collect::<Vec<_>>();
        let output = get_lang().utf_compose(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_reduce_reduced() {
        let input  = "foobar".chars().collect::<Vec<_>>();
        let output = get_lang().utf_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_reduce_nfc() {
        let input  = "foóbar".chars().collect::<Vec<_>>();
        let output = get_lang().utf_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_reduce_fill0() {
        let input  = "straße".chars().collect::<Vec<_>>();
        let output = get_lang().utf_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }
}