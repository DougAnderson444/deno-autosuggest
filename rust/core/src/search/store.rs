use crate::lexis::tokenize_query;
use super::data::{Record, SearchResult};
use super::engine::Engine;
use super::highlight::highlight;


static DEFAULT_LIMIT: usize = 10;


pub struct Store {
    records:    Vec<Record>,
    separators: (Vec<char>, Vec<char>),
    limit:      usize,
    results:    Vec<SearchResult>,
}


impl Store {
    pub fn new() -> Self {
        Self {
            records:    Vec::new(),
            separators: (vec!['['], vec![']']),
            limit:      DEFAULT_LIMIT,
            results:    Vec::with_capacity(DEFAULT_LIMIT),
        }
    }

    pub fn add(&mut self, record: Record) {
        self.records.push(record);
    }

    pub fn highlight_using(&mut self, separators: (&str, &str)) {
        let left:  Vec<char> = separators.0.chars().collect();
        let right: Vec<char> = separators.1.chars().collect();
        self.separators = (left, right);
    }

    pub fn results<'a>(&'a self) -> &'a [SearchResult] {
        &self.results
    }

    pub fn search(&mut self, query: &str) {
        let query: Vec<char> = query.chars().collect();
        let query = tokenize_query(&query);

        let mut engine = Engine::new(&query, self.limit);
        engine.push_many(&self.records);
        engine.sort_and_truncate();

        let separators = (&self.separators.0[..], &self.separators.1[..]);
        let results = &mut self.results;
        results.clear();
        results.extend(
            engine.hits()
                .iter()
                .map(|hit| {
                    SearchResult {
                        id: hit.id,
                        highlighted: highlight(hit, separators)
                    }
                })
        );
    }
}
