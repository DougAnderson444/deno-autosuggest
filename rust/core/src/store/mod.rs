use crate::lexis::{Text, tokenize_record};
use crate::lang::Lang;


static DEFAULT_LIMIT: usize = 10;


#[derive(Debug)]
pub struct Record {
    pub id:     usize,
    pub title:  Text<Vec<char>>,
    pub rating: usize,
}


impl Record {
    pub fn new(id: usize, source: &str, rating: usize, lang: &Option<Lang>) -> Record {
        Record {
            id,
            title: tokenize_record(source, lang),
            rating,
        }
    }
}


pub struct Store {
    pub records: Vec<Record>,
    pub limit:   usize,
    pub lang:    Option<Lang>,
    dividers:    (Vec<char>, Vec<char>),
}


impl Store {
    pub fn new() -> Self {
        Self {
            records:  Vec::new(),
            limit:    DEFAULT_LIMIT,
            lang:     None,
            dividers: (vec!['['], vec![']']),
        }
    }

    pub fn add(&mut self, record: Record) {
        self.records.push(record);
    }

    pub fn clear(&mut self) {
        self.records.clear();
    }

    pub fn highlight_with(&mut self, dividers: (&str, &str)) {
        let left:  Vec<char> = dividers.0.chars().collect();
        let right: Vec<char> = dividers.1.chars().collect();
        self.dividers = (left, right);
    }

    pub fn dividers<'a>(&'a self) -> (&'a [char], &'a [char]) {
        (&self.dividers.0, &self.dividers.1)
    }
}
