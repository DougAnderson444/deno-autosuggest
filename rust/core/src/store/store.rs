use std::cell::RefCell;
use crate::utils::to_vec;
use crate::lang::Lang;
use super::{Record, TrigramIndex, DEFAULT_LIMIT};


pub struct Store {
    pub next_ix:  usize,
    pub records:  Vec<Record>,
    pub limit:    usize,
    pub lang:     Lang,
    pub dividers: (Vec<char>, Vec<char>),
    pub index:    RefCell<TrigramIndex>,
    pub top_ixs:  RefCell<Option<Vec<usize>>>,
}


impl Store {
    pub fn new() -> Self {
        Self {
            next_ix:  0,
            records:  Vec::new(),
            limit:    DEFAULT_LIMIT,
            lang:     Lang::new(),
            dividers: (vec!['['], vec![']']),
            index:    RefCell::new(TrigramIndex::new()),
            top_ixs:  RefCell::new(None),
        }
    }

    pub fn add(&mut self, mut record: Record) {
        let Self { next_ix, index, records, .. } = self;
        debug_assert!(*next_ix == records.len(), "Invalid store.next_ix");
        record.ix = *next_ix;
        index.borrow_mut().add(&record);
        records.push(record);
        *next_ix += 1;
    }

    pub fn clear(&mut self) {
        self.records.clear();
        self.next_ix = 0;
    }

    pub fn highlight_with(&mut self, dividers: (&str, &str)) {
        let left:  Vec<char> = to_vec(dividers.0);
        let right: Vec<char> = to_vec(dividers.1);
        self.dividers = (left, right);
    }

    pub fn dividers<'a>(&'a self) -> (&'a [char], &'a [char]) {
        (&self.dividers.0, &self.dividers.1)
    }
}
