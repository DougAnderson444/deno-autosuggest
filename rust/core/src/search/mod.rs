mod score;
mod filter;
mod sort;
mod highlight;

use std::default::Default;
use crate::utils::LimitSortIterator;
use crate::tokenization::TextRef;
use crate::matching::WordMatch;
use crate::store::{Store, Record};


#[derive(Debug)]
pub struct Hit<'a> {
    pub id:      usize,
    pub title:   TextRef<'a>,
    pub rating:  usize,
    pub matches: Vec<WordMatch>,
    pub scores:  Scores,
}


impl<'a> Hit<'a> {
    pub fn from_record(record: &'a Record) -> Hit<'a> {
        Hit {
            id:       record.id,
            title:    record.title.to_ref(),
            rating:   record.rating,
            scores:   Default::default(),
            matches:  Vec::new(),
        }
    }
}


const SCORES_SIZE: usize = 9;


pub enum ScoreType {
    SameWords   = 0,
    SameNonFunc = 1,
    Typos       = 2,
    Trans       = 3,
    Fin         = 4,
    Offset      = 5,
    Rating      = 6,
    WordLen     = 7,
    CharLen     = 8,
}


#[derive(Debug, Clone)]
pub struct Scores([isize; SCORES_SIZE]);


impl Scores {
    pub fn iter(&self) -> impl Iterator<Item=&isize> {
        self.0.iter()
    }
}


impl std::ops::Index<ScoreType> for Scores {
    type Output = isize;

    fn index(&self, score: ScoreType) -> &Self::Output {
        &self.0[score as usize]
    }
}


impl std::ops::IndexMut<ScoreType> for Scores {
    fn index_mut(&mut self, score: ScoreType) -> &mut Self::Output {
        &mut self.0[score as usize]
    }
}


impl Default for Scores {
    fn default() -> Scores {
        Scores([0; SCORES_SIZE])
    }
}


#[derive(Debug)]
pub struct SearchResult {
    pub id:    usize,
    pub title: String,
}


impl Store {
    pub fn search<'a>(
        &'a self,
        query: &'a TextRef<'a>,
    ) -> Vec<SearchResult> {
        let dividers = self.dividers();

        let index = &mut *self.index.borrow_mut();
        index.prepare(&query, self.limit * 3);

        self.records.iter()
            .filter(|record| {
                query.words.len() == 0 || index.matches(&record)
            })
            .map(|record| {
                Hit::from_record(record)
            })
            .map(move |mut hit| {
                score::score(query, &mut hit);
                hit
            })
            .filter(move |hit| {
                filter::hit_matches(query, hit)
            })
            .limit_sort(self.limit, sort::compare_hits)
            .map(move |hit| {
                SearchResult {
                    id:    hit.id,
                    title: highlight::highlight(&hit, dividers),
                }
            })
            .collect()
    }
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::tokenization::tokenize_query;
    use crate::lang::{Lang, lang_english, lang_german};
    use crate::store::{Store, Record};

    fn check(name: &str, lang: Option<Lang>, queries: &[&str]) {
        let mut store = Store::new();
        store.lang = lang;
        store.add(Record::new(10, "brown plush bear",     10, &store.lang));
        store.add(Record::new(20, "the metal detector",   20, &store.lang));
        store.add(Record::new(30, "yellow metal mailbox", 30, &store.lang));
        store.add(Record::new(40, "thesaurus",            40, &store.lang));
        store.add(Record::new(50, "wi-fi router",         50, &store.lang));

        for (i, query) in queries.iter().enumerate() {
            let query   = tokenize_query(query, &store.lang);
            let query   = query.to_ref();
            let results = store.search(&query);
            assert_debug_snapshot!(format!("{}-{}", name, i), results);
        }
    }

    #[test]
    fn search_empty() {
        check("empty", None, &[""]);
    }

    #[test]
    fn search_equal() {
        check("equal", None, &["yelow metall maiblox"]);
    }

    #[test]
    fn search_partial() {
        check("partial", None, &[
            "brown plush bear",
            "metal detector",
            "yellow metal mailbox",
        ]);
    }

    #[test]
    fn search_intersection() {
        check("intersection", None, &[
            "red wooden mailbox",
            "red wooden mail",
        ]);
    }

    #[test]
    fn search_min_match() {
        check("min_match", None, &[
            "wooden mai",
            "wooden mail",
        ]);
    }

    #[test]
    fn search_transpositions() {
        check("transpositions", None, &[
            "metal mailbox",
            "mailbox metal",
        ]);
    }


    #[test]
    fn search_stemming() {
        let mut store = Store::new();
        store.lang = Some(lang_english());
        store.add(Record::new(30, "universe", 30, &store.lang));

        let query1   = tokenize_query("university", &None);
        let query2   = tokenize_query("university", &store.lang);
        let query1   = query1.to_ref();
        let query2   = query2.to_ref();
        let results1 = store.search(&query1);
        let results2 = store.search(&query2);

        assert_debug_snapshot!(results1);
        assert_debug_snapshot!(results2);
    }

    #[test]
    fn search_particles() {
        check("particles_nolang", None, &[
            "the",
        ]);

        check("particles", Some(lang_english()), &[
            "the",
        ]);
    }

    #[test]
    fn search_joined() {
        check("joined_query", None, &[
            "wifi",
        ]);

        check("joined_record", None, &[
            "the saurus",
        ]);
    }


    #[test]
    fn search_utf_normalization() {
        let mut store = Store::new();
        store.lang = Some(lang_german());
        store.add(Record::new(10, "Mitteltöner", 10, &store.lang));
        store.add(Record::new(20, "Passstraße",  20, &store.lang));

        let queries = [
            "mitteltö",
            "mitteltö", // ö in nfd!
            "mittelto",
            "passstras",
        ];

        for query in &queries {
            let query  = tokenize_query(query, &store.lang);
            let query  = query.to_ref();
            let result = store.search(&query);
            assert_debug_snapshot!(result);
        }
    }
}
