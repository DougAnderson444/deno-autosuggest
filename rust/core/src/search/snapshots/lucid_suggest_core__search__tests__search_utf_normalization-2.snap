---
source: src/search/mod.rs
expression: result
---
[
    SearchResult {
        id: 10,
        title: "[Mittelt√∂]ner",
    },
]
